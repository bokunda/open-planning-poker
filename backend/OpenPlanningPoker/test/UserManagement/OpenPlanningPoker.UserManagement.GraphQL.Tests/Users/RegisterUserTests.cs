namespace OpenPlanningPoker.UserManagement.GraphQL.Tests.Users;

public class RegisterUserTests
{
    private const string Mutation = """
            mutation RegisterUser($input: RegisterUserInput!) {
                registerUser(input: $input) {
                  user {
                    username
                    __typename
                  }
                  errors {
                    ... on OpenPlanningPokerError {
                      code
                      message
                      __typename
                    }
                  }
                }
            }
        """;

    private Dictionary<string, object?> GetVariables(string username) => new()
    {
        ["input"] = new Dictionary<string, object?>
        {
            ["username"] = username
        }
    };

    [Fact]
    public async Task RegisterUser_Success()
    {
        // arrange
        const string Username = "Username123";


        var usernameGeneratorService = Substitute.For<IUsernameGeneratorService>();
        usernameGeneratorService.GenerateUsername(Arg.Any<Language>(), Arg.Any<CancellationToken>())
            .Returns(Username);

        var executor = await new ServiceCollection()
            .AddSingleton(usernameGeneratorService)
            .AddGraphQL()
            .AddQueryConventions()
            .AddMutationConventions()
            .AddTypes()
            .BuildRequestExecutorAsync();

        var variables = GetVariables(Username);

        // act
        var result = await executor.ExecuteAsync(Mutation, variables);


        // assert
        result.MatchSnapshot();
    }

    [Fact]
    public async Task RegisterUser_NoUsername_Success()
    {
        // arrange
        const string Username = "";

        var usernameGeneratorService = Substitute.For<IUsernameGeneratorService>();
        usernameGeneratorService.GenerateUsername(Arg.Any<Language>(), Arg.Any<CancellationToken>())
            .Returns("Username123");

        var executor = await new ServiceCollection()
            .AddSingleton(usernameGeneratorService)
            .AddGraphQL()
            .AddQueryConventions()
            .AddMutationConventions()
            .AddTypes()
            .BuildRequestExecutorAsync();

        var variables = GetVariables(Username);

        // act
        var result = await executor.ExecuteAsync(Mutation, variables);


        // assert
        result.MatchSnapshot();
    }

    [Fact]
    public async Task RegisterUser_ToManyCharacters_Failed()
    {
        // arrange
        const string Username = "Username123Username123Username123Username123Username123Username123";

        var usernameGeneratorService = Substitute.For<IUsernameGeneratorService>();
        var executor = await new ServiceCollection()
            .AddSingleton(usernameGeneratorService)
            .AddGraphQL()
            .AddQueryConventions()
            .AddMutationConventions()
            .AddTypes()
            .BuildRequestExecutorAsync();

        var variables = GetVariables(Username);

        // act
        var result = await executor.ExecuteAsync(Mutation, variables);


        // assert
        result.MatchSnapshot();
    }
}
