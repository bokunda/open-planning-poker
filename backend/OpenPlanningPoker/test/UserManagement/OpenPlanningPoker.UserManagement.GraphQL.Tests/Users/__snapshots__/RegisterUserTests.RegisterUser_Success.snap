{
  "Kind": "SingleResult",
  "RequestIndex": null,
  "VariableIndex": null,
  "Label": null,
  "Path": null,
  "Data": {
    "registerUser": {
      "registerUserResponse": {
        "id": "e65fb783-3952-411c-b41c-1d5dc1d9ef6c",
        "userName": "Username123",
        "__typename": "RegisterUserResponse"
      },
      "errors": null,
      "__typename": "RegisterUserPayload"
    }
  },
  "Items": null,
  "Errors": null,
  "Extensions": null,
  "Incremental": null,
  "ContextData": null,
  "HasNext": null,
  "IsDataSet": true
}
